### ISSUE TITLE
When removing all layouts from a flexible, removeCallback from layouts are not called

### ISSUE THREAD

**Issue by NorthBlue333** [whitecube/nova-flexible-content#216]
Hi,
I am using Nova Advanced Medialibrary with your package, and I want to delete files not used anymore. I'm currently overriding the method `removeCallback` to remove those files, but actually the callback is not called if all layouts are removed !
Is this on purpose ?
See the file ` Whitecube\NovaFlexibleContent\Flexible.php`, in the method `syncAndFillGroups` : 

```php
    /**
     * Process an incoming POST Request
     *
     * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
     * @param  string  $requestAttribute
     * @return array
     */
    protected function syncAndFillGroups(NovaRequest $request, $requestAttribute)
    {
        if (!($raw = $this->extractValue($request, $requestAttribute))) {
            $this->groups = collect();
            return;
        }

        $callbacks = [];

        $new_groups  = collect($raw)->map(function ($item, $key) use ($request, &$callbacks) {
            $layout = $item['layout'];
            $key = $item['key'];
            $attributes = $item['attributes'];

            $group = $this->findGroup($key) ?? $this->newGroup($layout, $key);

            if (!$group) return;

            $scope = ScopedRequest::scopeFrom($request, $attributes, $key);
            $callbacks = array_merge($callbacks, $group->fill($scope));

            return $group;
        });

        $this->fireRemoveCallbacks($new_groups);

        $this->groups = $new_groups;

        return $callbacks;
    }
```

precisely
```php
        if (!($raw = $this->extractValue($request, $requestAttribute))) {
            $this->groups = collect();
            return;
        }
```

I am currently overriding this method as a workaround : 

```php
    /**
     * Process an incoming POST Request
     *
     * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
     * @param  string  $requestAttribute
     * @return array
     */
    protected function syncAndFillGroups(NovaRequest $request, $requestAttribute)
    {
        if (!($raw = $this->extractValue($request, $requestAttribute))) {
            $this->groups->each(function ($group) {
                if (method_exists($group, 'fireRemoveCallback')) {
                    $group->fireRemoveCallback($this);
                }
            });
            $this->groups = collect();
            return;
        }

        $callbacks = [];

        $new_groups  = collect($raw)->map(function ($item, $key) use ($request, &$callbacks) {
            $layout = $item['layout'];
            $key = $item['key'];
            $attributes = $item['attributes'];

            $group = $this->findGroup($key) ?? $this->newGroup($layout, $key);

            if (!$group) return;

            $scope = ScopedRequest::scopeFrom($request, $attributes, $key);
            $callbacks = array_merge($callbacks, $group->fill($scope));

            return $group;
        });

        $this->fireRemoveCallbacks($new_groups);

        $this->groups = $new_groups;

        return $callbacks;
    }
```

I could create a PR to fix this and to add some more usage to the `HasMediaLibrary` trait if you wish

**Comment by Nyratas** [comment]
Hi @NorthBlue333,

Thanks for reporting this issue. I did not test your workaround, but why don't you simply do:

```php
if (!($raw = $this->extractValue($request, $requestAttribute))) {
    $this->fireRemoveCallbacks(collect());
    $this->groups = collect();
    return;
}
```

It avoids code duplication and enhances code readability. Also, we could add some improvements to the `fireRemoveCallbacks` method in order to make it work without a `$new_groups` collection as parameter.

Of course, please make a PR for this, it is definitely a bug.

For the other suggested changes in `HasMediaLibrary`, can you open separated PRs so we don't have to mix subjects? Thanks!

Have a great day.

**Comment by NorthBlue333** [comment]
Thanks a lot for your fast answer! Just opened a PR for that.

I'll take a look for another one to update the `HasMediaLibrary` trait properly. I'm actually running in a segfault error right now and having a hard time figuring out where it comes from...

**Comment by Nyratas** [comment]
Closed in #217. Thanks @NorthBlue333 ;)
